Add support for the `theme()` function in the plugin API

adamwathan opened the pull request:
This PR adds support for the [`theme()` function](https://tailwindcss.com/docs/plugins#dynamic-utilities) from the v3 plugin API, used for configuring which values functional utilities support:

```js
plugin(function({ matchUtilities, theme }) {
  matchUtilities(
    {
      tab: (value) => ({
        tabSize: value
      }),
    },
    { values: theme('tabSize') }
  )
})
```

Things this handles:

- "Upgrading" theme keys to their v4 names, so if you do `theme('colors')` that will correctly retrieve all the colors from the `--color-*` namespace with the new CSS variable based configuration
- Polyfilling dependent keys, so `theme('backgroundColor')` will still pull everything in `--color-*` even though there is no values in the `backgroundColor` namespace in v4 by default
- Polyfilling theme values that are now handled by "bare values" internally, so even though there is no `flexShrink` theme values in v4, `theme('flexShrink')` will still configure your plugin to properly support any value that the built-in `shrink-*` utilities support

Things that aren't handled:

- Theme values that have been replaced by static utilities can't be retrieved yet, so for example `theme('cursor')` returns nothing right now because there are no values for the `cursor-*` utilities in the theme anymore, they are all just baked in to the framework.

This will be handled in a future PR.

adamwathan committed:
WIP

adamwathan committed:
Reorganize tests

adamwathan committed:
WIP

thecrypticace committed:
wip

thecrypticace committed:
wip

thecrypticace committed:
Cleanup types

thecrypticace committed:
Remove test

thecrypticace committed:
wip

thecrypticace committed:
wip

thecrypticace committed:
wip

thecrypticace committed:
wip

thecrypticace committed:
wip

thecrypticace committed:
wip

adamwathan committed:
WIP

thecrypticace committed:
Remove unncessary spread

thecrypticace committed:
Support fractional bare values

thecrypticace committed:
Add bare values test for all theme keys

thecrypticace committed:
Remove files

thecrypticace committed:
wip

adamwathan commented on packages/tailwindcss/src/ast.ts:
```diff
@@ -48,17 +48,17 @@ export function objectToAst(obj: CssInJs): AstNode[] {
         name = name.replace(/([A-Z])/g, '-$1').toLowerCase()
 
-        ast.push(decl(name, value))
+        ast.push(decl(name, String(value)))
```
I wonder if we could do this casting in the plugin API somewhere instead of here, so that core can always assume that values are strings?

thecrypticace replied:
`objectToAst` is only used by the plugin API so it still seems reasonable for it to be in that function. Perhaps we can move the fn though so it still feels self contained.

adamwathan replied:
Yep okay that's cool!

thecrypticace replied:
Done.

adamwathan commented on packages/tailwindcss/src/compat/config/resolve-config.ts:
```diff
@@ -0,0 +1,228 @@
+  plugins: [],
+}
+
+export function resolveConfig(configs: UserConfig[]): ResolvedConfig {
```
Should we simplify this for now to just be for resolving multiple themes rather than entire configs? Then use whatever we end up with here as part of resolving the whole config when we actually start working on support for JS config files? Feels wrong to me to add this stuff prematurely as part of this PR, even though of course it will eventually be needed when we actually do that work.

thecrypticace replied:
Done.

adamwathan commented on packages/tailwindcss/src/plugin-api.ts:
```diff
@@ -24,15 +30,23 @@ export type PluginAPI = {
 const IS_VALID_UTILITY_NAME = /^[a-z][a-zA-Z0-9/%._-]*$/
 
-export function buildPluginApi(designSystem: DesignSystem, ast: AstNode[]): PluginAPI {
+export function buildPluginApi(
```
Not really related to this PR but I _think_ we don't have to export this now that we are exporting `registerPlugins` instead 👍

thecrypticace replied:
Done.

thecrypticace committed:
Cleanup resolve config

wip

thecrypticace committed:
Remove unncessary export

thecrypticace committed:
Move `objectToAst`

RobinMalfait commented on packages/tailwindcss/src/compat/config/deep-freeze.ts:
```diff
@@ -0,0 +1,14 @@
+export function deepFreeze<T extends object>(obj: T): T {
```
I think this is unused?

RobinMalfait commented on packages/tailwindcss/src/compat/config/resolve-config.ts:
```diff
@@ -0,0 +1,141 @@
+      }
+
+      // The path contains a key that doesn't exist
+      if (!(key in value)) {
```
Would it be better to use `hasOwn` here? `hasOwn` only checks direct properties and not inherited once via the prototype chain.

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/hasOwn#direct_vs._inherited_properties

```suggestion
      if (!Object.hasOwn(value, key)) {
```

RobinMalfait commented on packages/tailwindcss/src/plugin-api.ts:
```diff
@@ -186,27 +212,400 @@ export function buildPluginApi(designSystem: DesignSystem, ast: AstNode[]): Plug
+      let result = resolvedConfig.theme ?? {}
+
+      // 1. Convert `path` to an array of parts (segment)
+      // 2. Find the "obeject" that corresponds to the path
```
```suggestion
      // 2. Find the "object" that corresponds to the path
```

RobinMalfait commented on packages/tailwindcss/src/plugin-api.ts:
```diff
@@ -186,27 +212,400 @@ export function buildPluginApi(designSystem: DesignSystem, ast: AstNode[]): Plug
+    theme(path) {
+      let cssValue = readFromCss(designSystem.theme, path)
+
+      if (typeof cssValue !== 'object') {
```
Should we check for `null` here?

thecrypticace replied:
Technically it's not necessary because `typeof null === "object"` _and_ we want that case to fall through. I've added an explicit check anyway + a clarifying comment just so the intent of the code is clearer.

RobinMalfait commented on packages/tailwindcss/src/plugin-api.ts:
```diff
@@ -186,27 +212,400 @@ export function buildPluginApi(designSystem: DesignSystem, ast: AstNode[]): Plug
+    }
+  }
+
+  // Now merge all the configs and make all that crap work
+  let resolvedConfig = resolveConfig([
+    createCompatabilityConfig(designSystem.theme),
+    ...pluginObjects.map(({ config }) => config ?? {}),
+  ])
+
+  let pluginApi = buildPluginApi(designSystem, ast, resolvedConfig)
```
Can we do this conditionally based on whether the legacy APIs were used? I think ideally we don't pay any costs if you don't use any legacy stuff.

Maybe a first step is to only do this if we have `pluginObjects` at all?

thecrypticace replied:
Yeah ideally we'd lazily create this only as necessary. I have some ideas but I think we need to consider its use by the CSS theme function before doing that.

thecrypticace replied:
We'll handle this in @philipp-spiess's CSS theme fn PR as that will likely inform the design.

RobinMalfait commented on packages/tailwindcss/src/plugin-api.ts:
```diff
@@ -186,27 +212,400 @@ export function buildPluginApi(designSystem: DesignSystem, ast: AstNode[]): Plug
+        return value.fraction
+      }),
+      backdropBlur: ({ theme }) => theme('blur'),
+      backdropBrightness: ({ theme }) => ({
+        ...theme('brightness'),
+        ...barePercentages,
+      }),
```
Oh nice solution regarding the `bareXXX`!

philipp-spiess commented on packages/tailwindcss/src/plugin-api.ts:
```diff
@@ -186,27 +212,400 @@ export function buildPluginApi(designSystem: DesignSystem, ast: AstNode[]): Plug
+    }
+  }
+
+  // Now merge all the configs and make all that crap work
+  let resolvedConfig = resolveConfig([
+    createCompatabilityConfig(designSystem.theme),
+    ...pluginObjects.map(({ config }) => config ?? {}),
+  ])
```
Nothing to add to this PR but I am wondering how we can make this accessible from where we evaluate the CSS `theme()` function. 🤔 Do you have any ideas?

We need the CSS theme function in two places right now:
- In a walk over all property values and `@theme` directives (currently we do our own walk and I think it's fine to do it after plugins are initialized since there are no downstream dependencies right now.
- In another walk on the AST created from utilities

Ideally we don't need to pass this legacy config around in the core of Tailwind. Maybe we can pass the Plugin API around. Another idea is to do the walk after we call register the plugins in this function but that won't allow us to hook it up when we parse the AST from utilities later.

thecrypticace replied:
We'll handle this in your CSS theme fn PR as that will likely inform the design. I have some ideas on how this will work but need to work through them to see if any of them are good or not.

philipp-spiess commented on packages/tailwindcss/src/plugin-api.ts:
```diff
@@ -186,27 +212,400 @@ export function buildPluginApi(designSystem: DesignSystem, ast: AstNode[]): Plug
+    | [main: string, extra: Record<string, string>]
+  // This only supports reading from the CSS Theme:…
+  let original = path
+    // Escape dots used inside square brackets
+    .replace(/\[(.*?)\]/g, (_, value) => `-${value.replace('.', '_')}`)
```
While on the CSS `theme()` function I learned that this is valid syntax in V3: `theme(color[red].[500])`

Here's how I handled it, we might want to do something similar here I suppose: https://github.com/tailwindlabs/tailwindcss/pull/14177/files#diff-3e0c118787a4d539fc94ae61398cd2b7117ce62372be52991f873946336d3aa6R101-R126

adamwathan replied:
@philipp-spiess Did you find usage like this in the wild or was it just something that was working in v3 perhaps by coincidence and not by design?

philipp-spiess replied:
@adamwathan I found some unit tests that documented this behavior in V3: https://github.com/tailwindlabs/tailwindcss/blob/main/tests/evaluateTailwindFunctions.test.js#L68-L93

Let me see if I can find something in the wild.

philipp-spiess replied:
Didn't find anything in the open that was using more than one square bracket block in a `theme()` call but there are some that use brackets as separators e.g. `colors['blue-gray'].100`

philipp-spiess replied:
Ok that example isn't from a tailwind codebase lol

adamwathan replied:
> @adamwathan I found some unit tests that documented this behavior in V3: https://github.com/tailwindlabs/tailwindcss/blob/main/tests/evaluateTailwindFunctions.test.js#L68-L93
> 
> Let me see if I can find something in the wild.

Hmm that doesn't look the same to me, I don't see a dot between the square brackets there right?

```css
.banana {
  color: theme('colors[yellow][100]');
}
```

I dont see `].[` anywhere in that test file that I can find.

thecrypticace committed:
Update packages/tailwindcss/src/plugin-api.ts

Co-authored-by: Robin Malfait <malfait.robin@gmail.com>

thecrypticace committed:
Remove unused function

thecrypticace committed:
Use `hasOwn` instead of `in`

thecrypticace committed:
Add explicit null check and clarifying comment

Technically`typeof cssValue !== 'object’` already covers this check _but_ that’s only because `typeof null == “object”` which is kinda absurd so an explicit check lets us be more clear about intent here

thecrypticace committed:
Handle more complex keypaths

This extracts a `toKeyPath` function that can be used by the CSS theme() function as well

thecrypticace committed:
Extract config to separate file

It’s gonna get larger eventually so might as well do it now

thecrypticace committed:
Unify `theme()` functions in plugins and `resolveConfig()`

thecrypticace committed:
Add test

thecrypticace committed:
Cleanup

thecrypticace committed:
Refactor

thecrypticace committed:
Add test

thecrypticace committed:
Cleanup

thecrypticace commented:
I think we should handle the static utilities case in a separate PR. Now that we have a "compat" config it's likely that we'll utilize that for the static utility case for plugins that need it — if it becomes necessary.

philipp-spiess commented on packages/tailwindcss/src/theme-fn.ts:
```diff
@@ -0,0 +1,152 @@
+    let configValue = resolveValue(get(result, keypath) ?? {})
+
+    if (typeof configValue === 'object') {
+      deepMerge(configValue, [cssValue], (_, b) => b)
+      set(result, keypath, configValue)
+
+      return get(result, keypath)
```
I am concerned that mutating the `configValue` from inside a function to _read_ from the config is going to create some unforeseen and hard-to-reason-about side effects (I couldn't come up with an issue right now but it does mean that calling `theme()` multiple times will update the plugin provided config over and over which seems really unexpected?). We don't need to do this, `return configValue` after the `deepMerge` also makes all tests pass.

I also saw that `deepMerge` mutates the input object though. Right now this is somewhat hidden because tests often use the function syntax for theme values so that every invocation would create a new object but for literal objects this could be an issue. Maybe we can instead create a new object here regardless and do something like this?

```suggestion
      return deepMerge({}, [configValue, cssValue], (_, b) => b)
```

thecrypticace replied:
Done

philipp-spiess commented on packages/tailwindcss/src/compat/config/create-compat-config.ts:
```diff
@@ -0,0 +1,179 @@
+export function createCompatConfig(theme: Theme): UserConfig {
+  return {
+    theme: {
+      colors: ({ theme }) => theme('color'),
```
So slick! 😍

philipp-spiess commented on packages/tailwindcss/src/compat/config/resolve-config.ts:
```diff
@@ -0,0 +1,121 @@
+    }
+  }
+
+  // Remove the `extend` key from the theme
+  // It's only used for merging and should
+  // not be present in the resolved theme
+  delete ctx.theme.extend
```
Nit but there seems to be some inconsistency when it comes to comment wrapping lengths.

philipp-spiess reviewed (commented):
LGTM but we should probably remove the mutation that happens with the `theme()` call.

philipp-spiess commented:
@thecrypticace Oh and the PR description could also use an update, it still says "Work in progress…" 🙈

philipp-spiess commented on packages/tailwindcss/src/theme-fn.ts:
```diff
@@ -0,0 +1,152 @@
+) {
+  return function theme(path: string) {
+    let keypath = toKeyPath(path)
+    let cssValue = readFromCss(designSystem.theme, keypath)
+
+    // If we get a single non-null value back from the CSS theme, return it
+    // Otherwise, we need to merge the CSS theme with the one resolved from
+    // all registered plugins.
+    if (cssValue === null || typeof cssValue !== 'object') {
+      return cssValue
+    }
```
Noticed one bug while rebasing the CSS `theme()` function. When we read form the CSS, we now no longer replace `colors` in the namespace with `color`. which means that `theme(colors.red.500)` won't work correctly.

thecrypticace replied:
yeep… I noticed this last night when working on some compatibility stuff. I think this means we might need to bring the explicit theme upgrade list back unfortunately. I'd prefer _not_ to if we can figure out a way though.

philipp-spiess replied:
This is even more annoying for things like `theme(backgroundColors.red.500)` where even if we replace `colors` to color`, it will still not be able to find the right CSS variable because we have no mapping to the CSS namespaces this should look into ( `--background-color` and `--color` in that order).

philipp-spiess replied:
@thecrypticace Yeah :/ Kinda like this right? https://github.com/tailwindlabs/tailwindcss/pull/14177/files#diff-3e0c118787a4d539fc94ae61398cd2b7117ce62372be52991f873946336d3aa6R214-R241

thecrypticace replied:
We (read: you) figured a way around this — thanks!

thecrypticace committed:
Merge consecutive keys during lookups

If we are looking for `backgroundColor.red.500` and `backgroundColor.red` does not exist then we’ll turn this into a lookup for `backgroundColor.red-500` which should (hopefully) exist because the keys are flattened in the CSS theme

thecrypticace committed:
Don’t mutate theme during lookup

thecrypticace committed:
Add test

thecrypticace committed:
Fix some bugs

thecrypticace committed:
Cleanup comments

thecrypticace committed:
Update changelog

thecrypticace committed:
Merge branch 'next' into feat/v4-plugin-theme-function
